.grit/patterns/moment-to-date-fns.grit

```diff
@@ -0,0 +1,765 @@
+
+    `$date.get($unit)` as $exp where {
+      $unit <: string(fragment=$unit_raw),
+      $unit_raw <: normalize_unit_specifier($norm),
+      $capitalized_unit = capitalize(string=$norm),
+      $duration_property = join(list=[$norm, "s"], separator=""),
+      $getfn = `get$capitalized_unit`,
+
+      if ($norm <: or { `millisecond`, `second`, `minute`, `hour` } ) {
+        $getfn = join(list=[$getfn, "s"], separator="")
+      },
+
+      $out = `((d => (d instanceof Date) ? datefns.$getfn(d) : (d.$duration_property ?? 0))($date))`,
+      if ($date <: identifier()) {
+        $out = `((d instanceof Date) ? datefns.$getfn(d) : (d.$duration_property ?? 0))`
+      },
+
+      if ($exp <: should_follow_semi()) {
+        $out = `;\n$out`
+      }
+    } => `$out`,
+
+    `$date.get($unit)` where { !$unit <: string() } => `getUnitFromDate($date, $unit)`,
+    `$date.set($unit, $value)` where { !$unit <: string() } => `setUnitOnDate($date, $unit, $value)`,
+
+    `$date.set($unit, $value)` as $exp where {
+      $unit <: string(fragment=$setter),
+      $setter <: or {
+        // units in a momentjs getter are case insensitive and allow plural forms
+        regex_year() => `setYear`,
+        regex_month() => `setMonth`,
+        regex_date() => `setDate`,
+        regex_hour() => `setHours`,
+        regex_minute() => `setMinutes`,
+        regex_second() => `setSeconds`,
+        regex_ms()  => `setMilliseconds`,
+      },
+      $comment = `/* TODO: date-fns objects are immutable, propagate this value appropriately */`,
+      $out = `datefns.$setter($date, $value)`,
+      if ($date <: identifier()) {
+        $out = `($date = datefns.$setter($date, $value))`
+      },
+      
+      if ($exp <: should_follow_semi()) {
+        $out = `;\n$out`
+      }
+    } => `$out`,
+  }
+}
+
+pattern rewrite_duration_set_and_get() {
+  `$duration.$getter()` where {
+    $getter <: or {
+      `asMilliseconds` => `toMilliseconds`,
+      `asSeconds` => `toSeconds`,
+      `asMinutes` => `toMinutes`,
+      `asHours` => `toHours`,
+      `asDays` => `toDays`,
+      `asWeeks` => `toWeeks`,
+      `asMonths` => `toMonths`,
+      `asYears` => `toYears`,
+    }
+  } => `durationfns.$getter($duration)`,
+}
+
+// Some stateful features of moment-js are simply missing in date-fns.
+// For example, calling `date.utc(<boolean>)` will set a flag on `date`
+// to use UTC to display a moment instead of the original moment's time.
+// This only affects the way a date is displayed (PST vs UTC).
+// This could technically be migrated by maintaining a global map of flags and
+// turning every `utc()` call into `set`, and every `display` call into a custom
+// overload, but I think that would introduce too much unwanted noise into the user's program.
+pattern missing_datefns_features() {
+  `$date.$method()` where {
+    $method <: or { `utc`, `local` }
+  } => `$date`
+}
+
+pattern rewrite_misc_methods() {
```

Comment:
Many of these are far too broad - ex. `$d.toJSON()` here will catch lots of *non-date* JSON objects. Let's make sure to tighten it up and add some negative test cases (code we shouldn't convert).

Reply:
It'll be impossible to always accurately determine the correct dispatch for `d` statically.
Instead, I think I'll do:

```js
(d instanceof Date ? ... : d.toJSON())
```

This way, all method calls matching `$d.toJSON` will be re-written, but the behavior won't change.

Reply:
It's still too broad. Between rewriting all toJSON instances throughout an app and potentially missing a few moment instances I lean heavily towards missing some moment instances.

Please update this so it only uses cases where $d was defined as a date.

Reply:
Understood. Though, I would've thought using a free function call like `toJSON_(foo)` would be clearer.
I'll change it to call the method instead.